man/mmime.1
.Dd July 22, 2016
.Dt MMIME 1
.Os
.Sh NAME
.Nm mmime
.Nd create MIME messages
.Sh SYNOPSIS
.Nm
.Op Fl c | Fl r
.Op Fl t Ar content-type
<
.Ar message
.Sh DESCRIPTION
.Nm
generates a
.Sq Li multipart/mixed
message from the standard input,
extending, wrapping, and encoding the header as necessary,
and replacing lines in the message body of the form
.Pp
.D1 Li # Ns Ar content Ns Li / Ns Ar type Ns Oo Ns Li # Ns Ar content-disposition Oc Pa path Ns Oo Li > Ns Ar filename Oc
.Pp
with a MIME part having Content-Type
.Ar content/type ,
consisting of the contents of the file found at
.Pa path .
.Ar content-disposition
is optional and defaults to
.Sq attachment .
.Ar filename
is optional and defaults to the basename of
.Ar path .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
Check mode: don't output anything,
exit with status 1 if MIME-encoding the message is required,
or else exit with status 0.
.It Fl r
Raw mode: don't expand MIME parts in the body, generate a
.Sq Li text/plain
message.
.It Fl t Ar content-type
Override Content-Type for the toplevel part.
Defaults to
.Sq Li multipart/mixed .
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev MBLAZE_RELAXED_MIME
If set,
.Nm Fl c
will relax the body line length check and only consider bodies with lines
longer than 998 characters need to be MIME formatted.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr makemime 1 ,
.Xr mhbuild 1
.Rs
.%A N. Freed
.%A N. Borenstein
.%D November 1996
.%R RFC 2045
.%T Multipurpose Internet Mail Extensions (MIME) Part One: Format of Internet Message Bodies
.Re
.Rs
.%A N. Freed
.%A N. Borenstein
.%D November 1996
.%R RFC 2046
.%T Multipurpose Internet Mail Extensions (MIME) Part Two: Media Types
.Re
.Rs
.%A K. Moore
.%D November 1996
.%R RFC 2047
.%T MIME (Multipurpose Internet Mail Extensions) Part Three: Message Header Extensions for Non-ASCII Text
.Re
.Rs
.%A N. Freed
.%A K. Moore
.%D November 1997
.%R RFC 2231
.%T MIME Parameter Value and Encoded Word Extensions: Character Sets, Languages, and Continuations
.Re
.Rs
.%A P. Resnick (ed.)
.%D October 2008
.%R RFC 5322
.%T Internet Message Format
.Re
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## 1.2

* Caution! The tools mdeliver and mexport were buggy in handling and
  generation of trailing empty lines in MBOX-RD.  Do not import
  mbox files generated by mexport >=1.2 with mdeliver <1.2 if you
  require verbatim message delivery.

## 1.1 (2021-01-14)

* mcom: allow tilde prefixed path for profile's outbox setting
* mcom: detect and report mmime errors
* add contrib/mmailto, a handler for mailto: links
* Bug fixes.

## 1.0 (2020-09-12)

* Caution! Backwards incompatible changes:
  * As a message name, `-` now refers to the message on the standard input,
    and not the previous message anymore.  Use `.-` to refer to the previous
    message in a short way.
    The tools will print a warning if you use `-` and the standard input
    comes from a TTY.
  * mpick: use the -F flag to read script files.
  * mpick: remove msglist support, use plain mmsg(7) arguments.
* Many mblaze tools now make use of pledge(2) on OpenBSD.
* add contrib/mcount, a tool to count mails
* mrep: use Reply-From configuration to find From header
* Many bug fixes.

## 0.7 (2020-05-15)

* All tools now follow symlinks to mails.
* mdirs: add -a to list all subfolders, ignoring Maildir++ convention.
* mcom: add preview alias for show command.
* mrep/mbnc: allow only one message as argument.
* maddr: add -d to only print display name.
* mthread: add -r to reverse top-level order.
* mlist: print number of matches when message selection is in place.
* mpick: many improvements.
* Many bug fixes.

## 0.6 (2020-01-18)

* mfwd: prefix Subject with "Fwd:".
* mscan: add dottime formatting.
* mlist: look at maildir/new too for messages.
* Many bug and portability fixes.

## 0.5.1 (2019-03-03)

* mdeliver: preserve mtime in mrefile
* mdirs: add -0 to separate folders by NUL characters
* Fixes for buffer-overflows, found by fuzzing.
* Fixes for memleaks.

## 0.5 (2019-02-09)

* New tool msearch to wrap several mail indexers.
* New zsh completion _mblaze.
* mnext/mprev were removed (you can call `mless +`/`mless .-`).
* The GnuPG tools in contrib/ now use gpg2.
* mshow exits with error if it could not extract all attachments
* mrep: add -noquote to disable quoting the message replied to
* mdeliver: keep permissions of messages
* mcom: aborting the editor is now more like delete than cancel
* mcom: add -send to send directly without editing
* mcom: check if mail is formatted sensibly
* mpick: new flag -v for statistics
* mscan: new flag -v for statistics
* magrep: add -h, which is like -p but doesn't print the file name
* mscan: prioritize displaying trashed mail over other markers
* mpick: fix off-by-one in expression parsing
* Many bug fixes.

## 0.4 (2018-08-15)

* New tool mrefile to move or copy messages.
* contrib/mp7m: add application/pkcs7-mime decoder
* mcom: allow setting arbitrary headers from command line
* mcom: add -body to prepopulate drafts
* mcom: mark drafts as seen after sending
* mcom: update flags after mrep/mbnc/mfwd
* mshow: add -B to decode MIME parameters in broken mails
* magrep: add -l flag
* Many bug fixes.

## 0.3.2 (2018-02-13)

* magrep: add *:REGEX to search in any header
* Fix of a buffer overflow in blaze822_multipart.
* Small bug fixes.
* Many documentation improvements by Larry Hynes.

## 0.3.1 (2018-01-30)

* mless: support $NO_COLOR
* mcolor: support $NO_COLOR
* blaze822.h: ensure PATH_MAX is defined
* Improved documentation.
* Many fixes for address parser.

## 0.3 (2018-01-12)

* New tool mflow to reformat format=flowed plain text mails.
* New tool mbnc to bounce mails (send original mail to someone else).
* mshow filters can output raw text now, e.g. for HTML rendering with colors.
* mrep can quote mail that doesn't have a plain text part.
* mcom runs mmime when deemed necessary.
* mhdr can extract MIME parameters.
* New contrib: mrecode
* New contrib: mraw
* mshow regards non-MIME mails as MIME mails with one part now.
* mshow -F to disable MIME filters.
* mpick supports negations now.
* msed can remove headers depending on their value.
* Improved UTF-8 parsing.
* Improved documentation.
* Numerous bug fixes and portability fixes.

## 0.2 (2017-07-17)

* New sequence syntax `m:+n` for `n` messages after message `m`.
* Threading shortcuts `=`, `_`, `^` for `.=`, `._`, `.^`.
* Sequence related errors are now reported.
* minc and mlist normalize slashes in paths.
* mfwd now generates conforming message/rfc822 parts.
* mthread can add optional folders (e.g. your outbox) to resolve message ids.
* mcom now adds Date: just before sending or cancelling the mail.
* VIOLATIONS.md documents how mblaze works with certain common mistakes.
* Full documentation revamp by Larry Hynes.
* Fix rare crash looking for mail body.
* Numerous small bug and portability fixes.

## 0.1 (2017-06-24)

* Initial release


mmime.c
#include <sys/stat.h>
#include <sys/types.h>

#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <pwd.h>
#include <search.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "blaze822.h"
#include "xpledge.h"

static int cflag;
static int rflag;
static char *tflag = "multipart/mixed";

int gen_b64(uint8_t *s, off_t size)
{
	static char *b64 =
	    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

	off_t i;
	int l;
	uint32_t v;
	for (i = 0, l = 0; i+2 < size; i += 3) {
		v = (s[i] << 16) | (s[i+1] << 8) | s[i+2];
		putc_unlocked(b64[(v & 0xfc0000) >> 18], stdout);
		putc_unlocked(b64[(v & 0x03f000) >> 12], stdout);
		putc_unlocked(b64[(v & 0x000fc0) >>  6], stdout);
		putc_unlocked(b64[(v & 0x3f)], stdout);
		l += 4;
		if (l > 72) {
			l = 0;
			printf("\n");
		}
	}
	if (size - i == 2) { // 2 bytes left, XXX=
		v = (s[size - 2] << 16) | (s[size - 1] << 8);
		putc_unlocked(b64[(v & 0xfc0000) >> 18], stdout);
		putc_unlocked(b64[(v & 0x03f000) >> 12], stdout);
		putc_unlocked(b64[(v & 0x000fc0) >>  6], stdout);
		putc_unlocked('=', stdout);
	} else if (size - i == 1) { // 1 byte left, XX==
		v = s[size - 1] << 16;
		putc_unlocked(b64[(v & 0xfc0000) >> 18], stdout);
		putc_unlocked(b64[(v & 0x03f000) >> 12], stdout);
		putc_unlocked('=', stdout);
		putc_unlocked('=', stdout);
	}
	printf("\n");
	return 0;
}

#define qphrasevalid(c) ((c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || \
                         (c >= 'a' && c <= 'z') || \
                         c == '!' || c == '*' || c == '+' || c == '-' || \
                         c == '/')

size_t
gen_qp(uint8_t *s, off_t size, size_t maxlinelen, size_t linelen)
{
	off_t i;
	int header = linelen > 0;
	char prev = 0;

	for (i = 0; i < size; i++) {
		// inspect utf8 sequence to not wrap in between multibyte
		int mb;
		if      ((s[i] & 0x80) == 0)    mb = 3;
		else if ((s[i] & 0xc0) == 0x80) mb = 3;
		else if ((s[i] & 0xe0) == 0xc0) mb = 6;
		else if ((s[i] & 0xf0) == 0xe0) mb = 9;
		else if ((s[i] & 0xf8) == 0xf0) mb = 12;
		else mb = 3;

		if (linelen >= maxlinelen-mb-!!header) {
			linelen = 0;
			prev = '\n';
			if (header) {
				printf("?=\n =?UTF-8?Q?");
				linelen += 11;
			} else {
				puts("=");
			}
		}

		if ((s[i] > 126) ||
		    (s[i] == '=') ||
		    (linelen == 0 &&
		     (strncmp((char *)s, "From ", 5) == 0 ||
	             (s[i] == '.' && i+1 < size &&
		      (s[i+1] == '\n' || s[i+1] == '\r'))))) {
			printf("=%02X", s[i]);
			linelen += 3;
			prev = s[i];
		} else if (header &&
			   (s[i] == '\n' || s[i] == '\t' || s[i] == '_')) {
			printf("=%02X", s[i]);
			linelen += 3;
			prev = '_';
		} else if (header && s[i] == ' ') {
			putc_unlocked('_', stdout);
			linelen++;
			prev = '_';
		} else if (s[i] < 33 && s[i] != '\n') {
			if ((s[i] == ' ' || s[i] == '\t') &&
			    i+1 < size &&
			    (s[i+1] != '\n' && s[i+1] != '\r')) {
				putc_unlocked(s[i], stdout);
				linelen += 1;
				prev = s[i];
			} else {
				printf("=%02X", s[i]);
				linelen += 3;
				prev = '_';
			}
		} else if (s[i] == '\n') {
			if (prev == ' ' || prev == '\t')
				puts("=");
			putc_unlocked('\n', stdout);
			linelen = 0;
			prev = 0;
		} else if (header && !qphrasevalid(s[i])) {
			printf("=%02X", s[i]);
			linelen += 3;
			prev = '_';
		} else {
			putc_unlocked(s[i], stdout);
			linelen++;
			prev = s[i];
		}
	}
	if (linelen > 0 && !header)
		puts("=");
	return linelen;
}

static const char *
basenam(const char *s)
{
	char *r = strrchr(s, '/');
	return r ? r + 1 : s;
}

static void
gen_attachment(const char *filename, char *content_disposition)
{
	const char *s = filename;
	int quote = 0;

	if (!*filename) {
		printf("Content-Disposition: %s\n", content_disposition);
		return;
	}

	for (s = (char *)filename; *s; s++) {
		if (*s < 32 || *s == '"' || *s >= 127 || s - filename > 35)
			goto rfc2231;
		if (strchr(" ()<>@,;:\\/[]?=", *s))
			quote = 1;
	}

	// filename SHOULD be an atom if possible
	printf("Content-Disposition: %s; filename=%s%s%s\n", content_disposition,
	    quote ? "\"" : "", filename, quote ? "\"" : "");
	return;

rfc2231:
	printf("Content-Disposition: %s", content_disposition);
	int i = 0;
	int d = 0;

	s = filename;

	while (*s) {
		i = printf(";\n filename*%d*=", d);
		if (d++ == 0) {
			printf("UTF-8''");
			i += 7;
		}
		while (*s && i < 78 - 3) {
			if (*s <= 32 || *s == '"' || *s > 126)
				i += printf("%%%02x", (uint8_t)*s++);
			else
				i += printf("%c", (uint8_t)*s++);
		}
	}

	printf("\n");
}

int
gen_file(char *file, char *ct)
{
	uint8_t *content;
	off_t size;

	char *cd = "attachment";
	char *s = strchr(ct, '#');
	if (s) {
		*s = 0;
		cd = s + 1;
	}

	const char *filename = basenam(file);
	s = strchr(file, '>');
	if (s) {
		*s = 0;
		filename = s + 1;
	}

	int r = slurp(file, (char **)&content, &size);
	if (r != 0) {
		fprintf(stderr, "mmime: error attaching file '%s': %s\n",
		    file, strerror(r));
		return -1;
	}

	if (strcmp(ct, "mblaze/raw") == 0)
		goto raw;

	off_t bithigh = 0;
	off_t bitlow = 0;
	off_t linelen = 0;
	off_t maxlinelen = 0;
	off_t i;
	for (i = 0; i < size; i++) {
		if (content[i] == '\n') {
			if (maxlinelen < linelen)
				maxlinelen = linelen;
			linelen = 0;
		} else {
			linelen++;
		}
		if (content[i] != '\t' && content[i] != '\n' && content[i] < 32)
			bitlow++;
		if (content[i] > 127)
			bithigh++;
	}

	gen_attachment(filename, cd);

	if (strcmp(ct, "message/rfc822") == 0) {
		printf("Content-Type: %s\n", ct);
		printf("Content-Transfer-Encoding: %dbit\n\n",
		    (bitlow > 0 || bithigh > 0) ? 8 : 7);
		fwrite(content, 1, size, stdout);
		return 0;
	}

	if (bitlow == 0 && bithigh == 0 &&
	    maxlinelen <= 78) {
		if (!ct)
			ct = "text/plain";
		printf("Content-Type: %s\n", ct);
		printf("Content-Transfer-Encoding: 7bit\n\n");
raw:
		fwrite(content, 1, size, stdout);
		return 0;
	} else if (bitlow == 0 && bithigh == 0) {
		if (!ct)
			ct = "text/plain";
		printf("Content-Type: %s\n", ct);
		printf("Content-Transfer-Encoding: quoted-printable\n\n");
		gen_qp(content, size, 78, 0);
		return 0;
	} else if (bitlow > size/10 || bithigh > size/4) {
		if (!ct)
			ct = "application/binary";
		printf("Content-Type: %s\n", ct);
		printf("Content-Transfer-Encoding: base64\n\n");
		return gen_b64(content, size);
	} else {
		if (!ct)
			ct = "text/plain";
		printf("Content-Type: %s\n", ct);
		printf("Content-Transfer-Encoding: quoted-printable\n\n");
		gen_qp(content, size, 78, 0);
		return 0;
	}
}

void
print_header(char *line) {
	char *s, *e;
	size_t l = strlen(line);

	if (l == 0)
		return;

	if (line[l-1] == '\n')
		line[l-1] = 0;

	/* iterate word-wise, encode words when needed. */

	s = line;

	if (!(*s == ' ' || *s == '\t')) {
		// raw header name
		while (*s && *s != ':')
			putc_unlocked(*s++, stdout);
		if (*s == ':')
			putc_unlocked(*s++, stdout);
	}

	int prevq = 0;  // was the previous word encoded as qp?
	char prevqs = 0;  // was the previous word a quoted-string?

	ssize_t linelen = s - line;

	while (*s) {
		size_t highbit = 0;
		int qs = 0;
		e = s;
		while (*e && *e == ' ')
			e++;

		if (*e == '"') {  // scan quoted-string, encode at once
			s = e;
			for (e++; *e && *e != '"'; e++) {
				if (*e == '\\')
					e++;
				if ((uint8_t)*e >= 127)
					highbit++;
			}
			if (*e == '"')
				e++;
			qs = 1;
		} else {  // scan word
			while (*e && *e == ' ')
				e++;
			for (; *e && *e != ' '; e++) {
				if ((uint8_t)*e >= 127)
					highbit++;
			}
		}

		if (!highbit) {
			if (e-s >= 998)
				goto force_qp;
			if (e-s >= 78 - linelen) {
				// wrap in advance before long word
				printf("\n");
				linelen = 0;
			}
			if (linelen <= 1 && s[0] == ' ' && s[1] == ' ') {
				// space at beginning of line
				goto force_qp;
			}
			if (*s != ' ' && !(prevqs && !prevq && *(s-1) != ' ')) {
				printf(" ");
				linelen++;
			}
			fwrite(s, 1, e-s, stdout);
			linelen += e-s;
			prevq = 0;
		} else {
force_qp:
			if (!prevq && *s == ' ')
				s++;

			if (qs && *s == '"')
				s++;
			if (qs && e > s && *(e-1) == '"')
				e--;

			if (linelen >= 78 - 13 - 4 ||
			    (e-s < (78 - 13)/3 &&
			     e-s >= (78 - linelen - 13)/3)) {
				// wrap in advance
				printf("\n");
				linelen = 0;
			}
			printf(" =?UTF-8?Q?");
			linelen += 11;
			linelen = gen_qp((uint8_t *)s, e-s, 78, linelen);
			printf("?=");
			linelen += 2;
			prevq = 1;

			if (qs && *e == '"')
				e++;
		}
		prevqs = qs;
		s = e;
	}
	printf("\n");
}

static int
valid_content_type(char *s)
{
	int slash = 0;

	for (; *s; s++)
		if (*s == '/')
			slash++;
		else if (isalnum(*s) || *s == '-' || *s == '+' || *s == '.' ||
		    *s == ';' || *s == '=' || *s == '#')
			; /* ok */
		else
			return 0;

	return slash == 1;
}

int
gen_build()
{
	char sep[100];
	snprintf(sep, sizeof sep, "----_=_%08lx%08lx%08lx_=_",
	    lrand48(), lrand48(), lrand48());

	char *line = 0;
	size_t linelen = 0;
	int inheader = 1;
	int intext = 0;
	int emptybody = 1;
	int ret = 0;
	char *contenttype = 0;
	char *contenttransferenc = 0;

	while (1) {
		ssize_t read = getdelim(&line, &linelen, '\n', stdin);
		if (read == -1) {
			if (feof(stdin)) {
				if (!emptybody)
					break;
				line = strdup(inheader ? "\n" : "");
			} else { // errored
				exit(1);
			}
		}
		if (inheader) {
			if (line[0] == '\n') {
				inheader = 0;
				printf("MIME-Version: 1.0\n");
				if (rflag) {
					printf("Content-Type:%s", contenttype ? contenttype : " text/plain; charset=UTF-8\n");
					printf("Content-Transfer-Encoding:%s", contenttransferenc ? contenttransferenc : " quoted-printable\n");
					printf("\n");
				} else {
					printf("Content-Type: %s; boundary=\"%s\"\n", tflag, sep);
					printf("\n");
					printf("This is a multipart message in MIME format.\n");
				}
			} else {
				if (strncasecmp(line, "Content-Type:", 13) == 0) {
					free(contenttype);
					contenttype = strdup(line+13);
				} else if (strncasecmp(line, "Content-Transfer-Encoding:", 26) == 0) {
					free(contenttransferenc);
					contenttransferenc = strdup(line+26);
				} else {
					print_header(line);
				}
			}
			continue;
		}

		if (!rflag && line[0] == '#') {
			char *f = strchr(line, ' ');
			if (f) {
				char of = *f;
				*f = 0;
				if (valid_content_type(line+1)) {
					printf("\n--%s\n", sep);
					if (line[read-1] == '\n')
						line[read-1] = 0;
					if (gen_file(f+1, line+1) != 0)
						ret = 1;
					intext = 0;
					emptybody = 0;
					continue;
				}
				*f = of;
			}
		}

		if (!rflag && !intext) {
			printf("\n--%s\n", sep);
			printf("Content-Type:%s", contenttype ? contenttype : " text/plain; charset=UTF-8\n");
			printf("Content-Disposition: inline\n");
			printf("Content-Transfer-Encoding:%s", contenttransferenc ? contenttransferenc : " quoted-printable\n");
			printf("\n");

			intext = 1;
		}

		if (contenttransferenc)
			printf("%s", line);
		else
			gen_qp((uint8_t *)line, strlen(line), 78, 0);

		emptybody = 0;
	}
	if (!rflag && !inheader)
		printf("\n--%s--\n", sep);

	free(line);
	return ret;
}

int
check()
{
	off_t bithigh = 0;
	off_t bitlow = 0;
	off_t linelen = 0;
	off_t maxheadlinelen = 0;
	off_t maxbodylinelen = 0;
	off_t bodylinelenlimit = getenv("MBLAZE_RELAXED_MIME") ? 998 : 78;

	int c;
	int l = -1;

	while ((c = getchar()) != EOF) {
		if (c == '\n') {
			if (maxheadlinelen < linelen)
				maxheadlinelen = linelen;
			linelen = 0;
			if (l == '\n')
				break;
		} else {
			linelen++;
		}
		if (c != '\t' && c != '\n' && c < 32)
			bitlow++;
		if (c > 127)
			bithigh++;
		l = c;
	}

	while ((c = getchar()) != EOF) {
		if (c == '\n') {
			if (maxbodylinelen < linelen)
				maxbodylinelen = linelen;
			linelen = 0;
		} else {
			linelen++;
		}
		if (c != '\t' && c != '\n' && c < 32)
			bitlow++;
		if (c > 127)
			bithigh++;
		l = c;
	}

	if (bitlow == 0 && bithigh == 0 &&
	    maxheadlinelen < 998 && maxbodylinelen <= bodylinelenlimit &&
	    l == '\n')
		return 0;
	else
		return 1;
}

int
main(int argc, char *argv[])
{
	srand48(time(0) ^ getpid());

	int c;
	while ((c = getopt(argc, argv, "crt:")) != -1)
		switch (c) {
		case 'r': rflag = 1; break;
		case 'c': cflag = 1; break;
		case 't': tflag = optarg; break;
		default:
usage:
			fprintf(stderr,
"Usage: mmime [-c|-r] [-t CONTENT-TYPE] < message\n");
			exit(1);
		}

	if (argc != optind)
		goto usage;

	xpledge("stdio rpath", "");

	if (cflag)
		return check();

	return gen_build();
}
